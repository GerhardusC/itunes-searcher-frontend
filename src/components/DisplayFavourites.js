//Redux imports
import { useSelector, useDispatch } from "react-redux"
import { removeFavourite } from "../redux/slices/favouritesSlice"

//This component displays a list of the user's favourites stored by Redux in state.
const DisplayFavourites = () => {
    //Get current favourites and define dispatch to remove a favourite.
    const currentFavourites = useSelector(state => state.favourites.data)
    const dispatch = useDispatch()
    //If favourites exist, return a list of the favourites in a table, otherwise return the text "No favourites to display."
    if(currentFavourites.length > 0){
        return (
            <div className="table-container">
                <table>
                    <thead>
                        <tr>
                            <th>Art</th>
                            <th>Name</th>
                            <th>Artist</th>
                            <th>Collection</th>
                        </tr>
                    </thead>
                    <tbody>
                        {
                            currentFavourites.map((favourite) => {
                                return (
                                    <tr key={favourite.trackViewUrl}>
                                        <td><img src={favourite.artworkUrl30 || favourite.artworkUrl60 || favourite.artworkUrl100} alt="Album art." /></td>
                                        <td>{favourite?.trackName || favourite.collectionName}</td>
                                        <td>{favourite.artistName}</td>
                                        <td>{favourite.collectionName}</td>
                                        <td>
                                            <button className="remove-button" onClick={() => dispatch(removeFavourite(favourite))}>‚ùå</button>
                                        </td>
                                    </tr>
                                )
                            })
                        }
                    </tbody>
                </table>
            </div>)
    } else {
        return (
            <div>No favourites to display</div>
        )
    }
}

export default DisplayFavourites;