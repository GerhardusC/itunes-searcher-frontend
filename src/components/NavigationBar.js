//Using global state to manage navigation instead of using Router and Routes.
import { useDispatch, useSelector } from "react-redux"
import { setViewFavourites, setViewSearch } from "../redux/slices/navigationStateSlice"
//The navigation bar component consists of buttons that allow you to change the current active page, which is the value of the navigation state.
const NavigationBar = () => {
    const dispatch = useDispatch()
    const navigationState = useSelector(state => state.navigation.value)
    return (
        <ul className="navigation-bar">
            <li><button onClick={() => dispatch(setViewSearch())} className={navigationState === "search" ? "active-nav-button" : "inactive-nav-button"}>Search</button></li>
            <li><button onClick={() => dispatch(setViewFavourites())} className={navigationState === "favourites" ? "active-nav-button" : "inactive-nav-button"}>Favourites</button></li>
        </ul>
    )
}

export default NavigationBar;