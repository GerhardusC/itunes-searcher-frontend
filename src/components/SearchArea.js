//Redux useDispatch is used to update the previous search term and to dispatch the api call with the search terms.
//It is also used to set the fetching status based on the api call's async status.
//React useState is used to keep track of user input within the component.
import { useDispatch, useSelector } from "react-redux";
import { update } from "../redux/slices/searchResultsSlice";
import { useState } from "react";
import {
  setFetchingStatus,
  setSearchTerm,
} from "../redux/slices/navigationStateSlice";

//The search area displays the search box for the search term, a selection box for media type and a button to search.
const SearchArea = () => {
  const dispatch = useDispatch();
  const [userSearchType, setUserSearchType] = useState("all");
  const [userSearchValue, setUserSearchValue] = useState("");
  const fetchingStatus = useSelector(
    (state) => state.navigation.fetchingStatus
  );

  //This function fetches results from our express backend with url queries.
  const getResults = async (term, type) => {
    let searchResults = {};
    //If the search term includes & symbol it messes with the api call, so we just disable users from being able to input it.
    if (!term.includes("&")) {
      try {
        dispatch(setFetchingStatus(true));
        let res = await fetch(
          `https://itunes-searcher-backend-production.up.railway.app/api?term=${term}&type=${type}`
        );
        searchResults = await res.json();
      } catch (err) {
        //If there is an error in the fetch we make a result that has a similar object structure to the normal results for convenience.
        searchResults = {
          resultCount: 0,
          results: {
            wrapperType: "error",
            statusText: "Problem fetching results from the server.",
          },
        };
        //If something goes wrong reset the search term.
        dispatch(setSearchTerm(""));
      } finally {
        //Clean up
        dispatch(setFetchingStatus(false));
        dispatch(setSearchTerm(userSearchValue));
        dispatch(update(searchResults));
      }
    }
    //Send an error object if the search term includes the & character.
    if (term.includes("&")) {
      searchResults = {
        resultCount: 0,
        results: {
          wrapperType: "error",
          statusText:
            "Invalid input. Make sure your search term does not include the symbol '&'",
        },
      };
      dispatch(setSearchTerm(""));
      dispatch(update(searchResults));
    }
  };

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        //This prevents the user from fetching again with the form submission before the fetch is completed.
        if (!fetchingStatus) {
          getResults(userSearchValue, userSearchType);
          setUserSearchValue("");
        }
      }}
    >
      <div>Enter search term:</div>
      <input
        placeholder="search"
        required
        onChange={(e) => setUserSearchValue(e.target.value)}
        value={userSearchValue}
      ></input>
      <div>Select category:</div>
      <select
        onChange={(e) => {
          setUserSearchType(e.target.value);
        }}
      >
        <option value="all">all</option>
        <option value="music">music</option>
        <option value="movie">movie</option>
        <option value="podcast">podcast</option>
        <option value="audiobook">audio book</option>
        <option value="software">software</option>
        <option value="ebook">ebook</option>
        <option value="tvShow">tv show</option>
        <option value="shortFilm">short film</option>
        <option value="musicVideo">music video</option>
      </select>
      <button type="submit" className="search-button">
        Search
      </button>
    </form>
  );
};

export default SearchArea;
