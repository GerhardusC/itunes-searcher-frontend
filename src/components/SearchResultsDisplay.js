//Using Redux and storing the search results in global state so we can add some of them to favourites.
import { useSelector, useDispatch } from "react-redux";
import { addFavourite } from "../redux/slices/favouritesSlice";

//This component displays the search results in a table based on the user search input submitted from the search area component.
const SearchResultsDisplay = () => {
  const searchResults = useSelector((state) => state.search.data);
  const fetchingStatus = useSelector(
    (state) => state.navigation.fetchingStatus
  );
  const searchTerm = useSelector((state) => state.navigation.searchTerm);
  const dispatch = useDispatch();
  //If data is being fetched return fetching data.
  if (fetchingStatus) {
    return <div>Fetching data...</div>;
  }
  //If search results are returned and the resultCount property says that there are results related to the search term,
  //then return a table displaying the results.

  if (searchResults && searchResults.resultCount > 0) {
    const searchIds = Object.keys(searchResults.results);

    return (
      <div>
        <p>
          Displaying {searchResults.resultCount} results for <b>{searchTerm}</b>
          .
        </p>
        <div className="table-container">
          <table>
            <thead>
              <tr>
                <th>Art</th>
                <th>Name</th>
                <th>Artist</th>
                <th>Collection</th>
              </tr>
            </thead>
            <tbody>
              {searchIds.map((key) => {
                return (
                  <tr key={key}>
                    {/* 
                                                Depending on which artwork exists, we display it in the table, some results only have 30px and some have only 60 px etc.
                                            */}
                    <td>
                      <img
                        src={
                          searchResults.results[key].artworkUrl30 ||
                          searchResults.results[key].artworkUrl60 ||
                          searchResults.results[key].artworkUrl100
                        }
                        alt="art"
                      />
                    </td>
                    {/* 
                                                If the result is an audiobook, it has no trackname, so we use the collectionName instead of the track name, because an audio book is classified
                                                as a collection according to the search api. 
                                            */}
                    <td>
                      {searchResults.results[key]?.trackName ||
                        searchResults.results[key].collectionName}
                    </td>
                    <td>{searchResults.results[key].artistName}</td>
                    <td>{searchResults.results[key].collectionName}</td>
                    <td>
                      {/* 
                                                    If the item is not currently favourited we give a favourite button, otherwise we return an empty.
                                                    All collectionIds in the currentIds array belong to audio books, and all the trackIds belong to everything else.
                                                */}
                      <button
                        className="favourite-button"
                        onClick={(e) => {
                          dispatch(addFavourite(searchResults.results[key]));
                          e.target.style.display = "none";
                        }}
                      >
                        ‚≠ê
                      </button>
                    </td>
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      </div>
    );
  }
  //There are two types of error objects that can be returned from the front end ofthe application if either the result contains the & symbol or
  //if the api call was unsuccessful. The statusText returned from this object is then displayed to the user.
  if (
    searchResults.resultCount === 0 &&
    searchResults.results.wrapperType === "error"
  ) {
    return (
      <>
        <div>{searchResults.results.statusText}</div>
      </>
    );
  }
  //If there are no results, but the result returned was not an error, display that no results were found from the store.
  if (searchResults.resultCount === 0) {
    return (
      <div>
        No results found for <b>{searchTerm}</b>.
      </div>
    );
  }
};

export default SearchResultsDisplay;
