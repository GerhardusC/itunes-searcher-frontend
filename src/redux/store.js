import { configureStore } from '@reduxjs/toolkit'
import searchReducer from './slices/searchResultsSlice'
import favouritesSlice from './slices/favouritesSlice'
import navigationStateSlice from './slices/navigationStateSlice'

//Creating Redux store for search results.
export const store = configureStore({
    reducer: {
        search: searchReducer,
        favourites: favouritesSlice,
        navigation: navigationStateSlice
    },
  })