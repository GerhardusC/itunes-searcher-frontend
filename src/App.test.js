import React from "react";
import NavigationBar from "./components/NavigationBar";
import DisplayFavourites from "./components/DisplayFavourites";
import SearchResultsDisplay from "./components/SearchResultsDisplay";
import SearchArea from "./components/SearchArea";
import renderer from "react-test-renderer";
import { store } from "./redux/store";
import { Provider } from "react-redux";

test("Navigation bar renders correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <NavigationBar />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Favourites display correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <DisplayFavourites />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Search results display correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <SearchResultsDisplay />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Search area display correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <SearchArea />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
