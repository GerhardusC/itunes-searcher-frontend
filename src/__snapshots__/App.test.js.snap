// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favourites display correctly 1`] = `
<div>
  No favourites to display
</div>
`;

exports[`Navigation bar renders correctly 1`] = `
<ul
  className="navigation-bar"
>
  <li>
    <button
      className="active-nav-button"
      onClick={[Function]}
    >
      Search
    </button>
  </li>
  <li>
    <button
      className="inactive-nav-button"
      onClick={[Function]}
    >
      Favourites
    </button>
  </li>
</ul>
`;

exports[`Search area display correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    Enter search term:
  </div>
  <input
    onChange={[Function]}
    placeholder="search"
    required={true}
    value=""
  />
  <div>
    Select category:
  </div>
  <select
    onChange={[Function]}
  >
    <option
      value="all"
    >
      all
    </option>
    <option
      value="music"
    >
      music
    </option>
    <option
      value="movie"
    >
      movie
    </option>
    <option
      value="podcast"
    >
      podcast
    </option>
    <option
      value="audiobook"
    >
      audio book
    </option>
    <option
      value="software"
    >
      software
    </option>
    <option
      value="ebook"
    >
      ebook
    </option>
    <option
      value="tvShow"
    >
      tv show
    </option>
    <option
      value="shortFilm"
    >
      short film
    </option>
    <option
      value="musicVideo"
    >
      music video
    </option>
  </select>
  <button
    className="search-button"
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Search results display correctly 1`] = `null`;
